dbms/src/Functions/FunctionsJson.h

```diff
@@ -530,11 +612,723 @@ class FunctionsJsonArray : public IFunction
+                        data_to,
+                        offsets_to,
+                        input_source,
+                        column_nullable.getNullMapData(),
+                        input_tidb_tp->tp(),
+                        input_tidb_tp->flen(),
+                        block.rows());
+                }
+                else
+                {
+                    doExecuteForBinary<true, false>(
+                        data_to,
+                        offsets_to,
+                        input_source,
+                        tmp_null_map->getData(),
+                        input_tidb_tp->tp(),
+                        input_tidb_tp->flen(),
+                        block.rows());
+                }
+            }
+            else
+            {
+                doExecuteForBinary<false, false>(
+                    data_to,
+                    offsets_to,
+                    input_source,
+                    tmp_null_map->getData(),
+                    input_tidb_tp->tp(),
+                    input_tidb_tp->flen(),
+                    block.rows());
+            }
+        }
+        else if ((unlikely(output_tidb_tp == nullptr)) || hasParseToJSONFlag(*output_tidb_tp))
+        {
+            if (from.column->isColumnNullable())
+            {
+                const auto & column_nullable = static_cast<const ColumnNullable &>(*from.column);
+                doExecuteForParsingJson<true>(
+                    data_to,
+                    offsets_to,
+                    input_source,
+                    column_nullable.getNullMapData(),
+                    block.rows());
+            }
+            else
+            {
+                auto tmp_null_map = ColumnUInt8::create(0, 0);
+                doExecuteForParsingJson<false>(
+                    data_to,
+                    offsets_to,
+                    input_source,
+                    tmp_null_map->getData(),
+                    block.rows());
+            }
+        }
+        else
+        {
+            doExecuteForOthers(data_to, offsets_to, input_source, block.rows());
+        }
+
+        if (from.column->isColumnNullable())
+        {
+            auto null_map = static_cast<const ColumnNullable &>(*from.column).getNullMapColumnPtr();
+            block.getByPosition(result).column = ColumnNullable::create(std::move(col_to), std::move(null_map));
+        }
+        else
+        {
+            block.getByPosition(result).column = std::move(col_to);
+        }
+    }
+
+private:
+    template <bool is_binary_str, bool check_null_for_binary_str>
+    static void doExecuteForBinary(
+        ColumnString::Chars_t & data_to,
+        ColumnString::Offsets & offsets_to,
+        const std::unique_ptr<IStringSource> & data_from,
+        const NullMap & null_map_from,
+        UInt8 from_type_code,
+        Int32 flen,
+        size_t size)
+    {
+        // json_type + from_type_code + string end char 0
```

Comment:
The string end char 0 should already be counted in `data_from->getSizeForReserve()` since the input column is a string column?

Reply:
yes, you are right, changed.